_posts/2020-06-18-life-locks.md

```diff
@@ -0,0 +1,155 @@
+
+```yaml
+- Sid: 'RequireObjectLockMode'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'Null':
+        's3:object-lock-mode': 'true'
+- Sid: 'RequireObjectLockModeToBeCompliance'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'StringNotEqualsIgnoreCase':
+        's3:object-lock-mode': 'COMPLIANCE'
+```
+
+#### s3:object-lock-remaining-retention-days
+
+Next you enforce how long uploaded files are protected by using the `s3:object-lock-remaining-retention-days` condition key. In the example below, you require that a retention period be specified, and that it must be 7 days. It is also possible to use this condition key to enforce a day range; I'll leave that as an exercise for you dear reader :wink:
+
+```yaml
+- Sid: 'RequireRetentionDays'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'Null':
+        's3:object-lock-remaining-retention-days': 'true'
+- Sid: 'RequireRetentionToBeThreeDays'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      NumericNotEquals:
+        's3:object-lock-remaining-retention-days': '7'
+```
+
+### Unspoken Requirement
+
+The last and unspoken requirement can be found in this [complete CloudFormation template](/assets/2020-06-03-life-locks-example-template.yml).
+
+## Definitions
+
+During your reading you realise that some of the terms in [Object Locking](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock.html) are a bit vague, so you document them for the rest of the team:
+
+* **LegalHold**: view this as making an object 'read-only'. It is independent to the `LockingMode`; you can specify neither, either or both of Legal Hold and Locking Mode.
+* **LockingMode**: view this as how tamper proof the WORM model is, and has one of the following values:
+  * **Governance**: Governance mode can be disabled by either an IAM User with the required IAM permissions or by the AWS Account's root user.
+  * **Compliance**: Compliance mode cannot be disabled by any user, including the Account's root user.
+
+**Note:** The `LockingMode` expiry date can only be modified if it extends the lock.
+
+## Fame, Glory and Accolades
+
+Phew, what a day! What started out with a shock quickly turned out very well! You managed to not only address the Compliance Manager's concerns but also have a proof-of-concept ready to deploy! Your colleagues are amazed, your manager is astounded and everyone else is seriously impressed by your Cloud Skillz!
+
+Well done!
```

Comment:
Good read. Perhaps just sign off with something more, ie. the usual hook into CNS or something